oak/server/rust/oak_runtime/src/node/external.rs

```diff
@@ -0,0 +1,97 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::runtime::{Handle, RuntimeProxy};
+use crate::NodeId;
+use lazy_static::lazy_static;
+use log::{error, info};
+use oak_abi::OakStatus;
+use std::sync::RwLock;
+use std::thread;
+use std::thread::JoinHandle;
+
+pub type NodeFactory = fn(config_name: &str, node_id: NodeId, handle: Handle);
+
+lazy_static! {
+    static ref FACTORY: RwLock<Option<NodeFactory>> = RwLock::new(None);
+}
+
+pub fn register_factory(factory: NodeFactory) {
+    info!("register external node factory");
+    *FACTORY.write().unwrap() = Some(factory);
+}
+
+pub struct PseudoNode {
+    config_name: String,
+    runtime: RuntimeProxy,
+    reader: Handle,
+    thread_handle: Option<JoinHandle<()>>,
+}
+
+impl PseudoNode {
+    pub fn new(config_name: &str, runtime: RuntimeProxy, reader: Handle) -> Self {
+        Self {
+            config_name: config_name.to_string(),
+            runtime,
+            reader,
+            thread_handle: None,
+        }
+    }
+}
+
+impl super::Node for PseudoNode {
+    fn start(&mut self) -> Result<(), OakStatus> {
+        // Clone or copy all the captured values and move them into the closure, for simplicity.
+        let config_name = self.config_name.clone();
+        let runtime = self.runtime.clone();
+        let reader = self.reader;
+        let thread_handle = thread::Builder::new()
+            .name(format!("external={}", config_name))
+            .spawn(move || {
+                let pretty_name = format!("{}-{:?}:", config_name, thread::current());
+                let factory_fn: NodeFactory = FACTORY
+                    .read()
+                    .expect("unlock failed")
+                    .expect("no registered factory");
+                info!(
```

Comment:
I'm not a big fan of all this two-way registering protocol, and also the fact that the Runtime does not actually know what's going to happen when creating an unrecognized pseudo-node. I think it's fine to assume that if we add a new pseudo-node type in C++, we can manage adding a bit of code to the Rust runtime to connect things in a more idiomatic way, and avoid the `External` configuration enum type too. Perhaps just keep an `ExternalNode` struct here that wraps the thread creation, and forwards to the actual C method that executes the main node event loop. WDYT?

Reply:
Yeah, there's a certain amount of path-of-least-resistance:
 - With `main` in C++, we only want explicit linkage from C++ to Rust, not vice versa (to avoid cycles)
 - Rust needs to up-call to C++ for pseudo-Node creation, which therefore has to happen via a function pointer.
 - Any such function pointer then has to be registered via a call from C++ to Rust.
 - A single such function pointer can currently be used for all pseudo-Node types, because the C++ runtime has a full copy of the configuration and so the config name is all that's needed, regardless of pseudo-Node type (no need to pass pseudo-Node-specific configuration values extracted from the `ApplicationConfiguration`).

A lot of the above will change once `main` is in Rust, so I don't think it's worth changing things here &ndash; #724 will require a re-work of the C++ pseudo-Node registration stuff anyway (and I've added a reminder `TODO`).